simulator/func_sim/func_sim.cpp

```diff
@@ -75,6 +75,59 @@ void FuncSim<ISA>::run( const std::string& tr, uint64 instrs_to_run)
     }
 }
 
+template <typename ISA>
+void FuncSim<ISA>::gdb_load( const std::string &tr) {
+    mem->load_elf_file( tr);
+    sout << "MIPT-MIPS: Binary file " << tr << " loaded" << std::endl;
+    //TODO: print more useful info
+}
+
+template <typename ISA>
+void FuncSim<ISA>::gdb_prepare() {
+    PC = mem->startPC();
+    nops_in_a_row = 0;
+    sout << "MIPT-MIPS: prepared to run" << std::endl;
+    //TODO: print more useful info
+}
+
+template <typename ISA>
+void FuncSim<ISA>::gdb_resume( int steps) {
+    sout << "MIPT-MIPS: resuming, steps: " << steps << std::endl;
+    uint32 instrs_to_run = ((steps == 0) ? MAX_VAL32 : steps);
+    for ( uint32 i = 0; i < instrs_to_run; ++i) {
+        const auto& instr = step();
+        sout << instr << std::endl;
+        if ( instr.is_halt())
+            break;
+    }
+}
+
+template <typename ISA>
+int FuncSim<ISA>::gdb_mem_read (unsigned int addr, unsigned char *buf, int length) try {
+    int bytes_read = 0;
+    while ( length-- > 0) {
+        *buf++ = static_cast<unsigned char>( mem->_read_byte ( addr++));
+        bytes_read++;
+    }
+    return bytes_read;
+}
+catch (...) {
+    return 0;
+}
+
+template <typename ISA>
+int FuncSim<ISA>::gdb_mem_write (unsigned int addr, const unsigned char *buf, int length) try {
+    int bytes_written = 0;
+    while ( length-- > 0) {
+        mem->_write_byte( addr++, *buf++);
+        bytes_written++;
+    }
+    return bytes_written;
+}
+catch (...) {
+    return 0;
```

Comment:
Should it return `bytes_written` instead?

Reply:
Yes, it should. I'll fix it.

Reply:
Then it contradicts with:
> GDB traits 0 as an error for this functions

Reply:
It's not clear from function description:
>Result is number of bytes write, or zero if error.

On the one hand, 0 is an error code. On the other hand, if we wrote/read less bytes than GDB reqested, then some error has happened. I'll try to look how GDB would treat both cases.